### DISCUSSION TITLE
API Design of how to use builtin plugins

### DISCUSSION THREAD

**Discussion by hyfdev** [rolldown/rolldown#1807] [Core Minutes]
overview

```js
import { defineConfig, importMetaVarsPlugin } from 'rolldown'
// or
import { importMetaVarsPlugin } from 'rolldown/plugins'


export default defineConfig({
  input: {
    entry: './index.ts',
  },
  plugins: [
    // No new XXXXPlugin
    new importMetaVarsPlugin({
      xx: true,
      fss: {},
      fn: () => {},
    }),
    // Supported by default
    importMetaVarsPlugin({
      xx: true,
      fss: {},
      fn: () => {},
    }),
    // ---------------
    // Enhancement A
    {
      builtin: 'import-meta-vars',
      xx: true,
      fss: {},
      fn: () => {},
    },
    // Enhancement B
    {
      name: 'builtin:import-meta-vars',
      xx: true,
      fss: {},
      fn: () => {},
    },
    // Enhancement C
    {
      name: 'builtin:import-meta-vars',
      options: {
        xx: true,
        fss: {},
        fn: () => {},
      },
    },
    // Enhancement D
    {
      builtin: 'import-meta-vars',
      options: {
        xx: true,
        fss: {},
        fn: () => {},
      },
    },
  ],
})

```

keys:

- Should we have an enhancement pattern that doesn't require `import { xxxPlugin }`? Does it provide better DX?
- Should we limit the plugins under the `rolldown/plugins`?

Notices:

Plugins that exported by `rolldown` are all native plugin writting in rust. For official js plugins, we will put them into different packages and have `@rolldown/plugin-xxx` names.

**Comment by 7086cmd** [comment]
I believe the default function initialization is sufficient. However, if an enhanced style is necessary, I support option **B**, as its configuration keys may be consistent with those of other plugins.

**Comment by Boshen** [comment]
```
    importMetaVarsPlugin({
      xx: true,
      fss: {},
      fn: () => {},
    }),
```

This should provide the best TypeScript typings support I think.

**Reply by hyfdev** [reply]
Yeah. Totally. Typings support will also come to the enhancement pattern if it get accepted.

**Comment by ikkz** [comment]
My preference is B. some reasons:
- no imports no news, very simple
- has the same `name` and options level with rollup's [official plugins](https://github.com/rollup/plugins)
- typing support is possible, [playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAQgrgSwDbAQOwApLgc3QeTFQHs0BnKAXigG8AoKRqAWwEM0EAzCM4ALloMmwgE7Fi-KLxHocAGiHDGxOMAytgACwHTZC4QF99TAOQ4kxAEYBaBMzDERwEwPpKmYiQH4dwGWnlFJREeYEcIAEU4CBEQAFEAD2AIcgRSAUtxJAh2IIM6fNBIWEQUdCxcdAAeABUoCCSUgBMKAGsIEGJOEuRUTGw8NEIScgA+KkFhNFZmCAEAA0tSvr4AEhoag3njZSI08gF4XvKBgj3SMgBtGoBdAro6Iugjsv7K8gm3JkuAaSh0KDtTrdF59CqDYb7Mg3Q7LE7vKo-UYFS5Aro9V7gs4jaEAbgeTygAFUyDEsWhPkFprNfP4cPjhBJNDEAGqsbA8ACMAnYIHx+Ue4GgACViEgkE1iAB3N6DCag+GDCgAH2JpJE5PxdHQyREnFYAGNoJCLpMmGBTgcoKLxZKZeTLncBZw4GgDSMoE0IJx0BAAMKkH04AAUBsDCBwAhN5AAlLQoAKvT60P7wyGvowLe8yAJLkFhAB6AtQTSsChFWRSOBgBxOOQsDQGzRQQlJ30cD3sJpQNDEeoiMQiMj5pgZ9xMalzKAmJbHNB8cxWWz2RzOHbj3Y41wjjeMTySABEzPFxAP693HlC4SiMXijVS6Sg+qQpJ3SnyG6Mb7HG8nAhncLzmwHDcLwJjnuOxDnFaP4XlARaesQPBoCYwANsATb1l2UiaNK-aDsOcHBOIkh+NEEEXioagaNo07HhYUqOBKJhvoYb5fhuCG9mg1izq8UBZkq36sROMxTiYcDqgJloIOBInKForLstEZDci2IjROxQROjGdBAA)
